### PR TITLE
Integrate weight-only quantizaion of INC

### PR DESCRIPTION
**Pull request by mengniwang95**
This PR integrate weight-only quantization of neural compressor into optimum-intel.

Notice: Need to use the master branch for test

### CODE REVIEW

**Comment by hshen14** [comment]
@echarlaix Could you please help review this PR? INC supports production-level quality of weight-only quantization including INT8 and INT4 for LLMs in latest master (also be released in INC v2.3 in early Sep). Thanks.

File: examples/neural_compressor/language-modeling/run_clm.py
Comment (echarlaix) [inline examples/neural_compressor/language-modeling/run_clm.py:732]: is it because GPTQ methology requires a lot of memory resource ? Can you also add a comment to make it clear

File: examples/neural_compressor/language-modeling/run_clm.py
Comment (echarlaix) [inline examples/neural_compressor/language-modeling/run_clm.py:626]: What happens if `quantization_approach == "weight_only"` and  `op_type_dict`  is not provided ? Should this be also added before quantization in the quantizer when not provided ?

File: examples/neural_compressor/language-modeling/run_clm.py
Comment (echarlaix) [inline examples/neural_compressor/language-modeling/run_clm.py:719]: Why is it used for ? Can we include it in the config or quantizer directly instead of modifying the original model ?

File: optimum/intel/neural_compressor/quantization.py
Comment (echarlaix) [inline optimum/intel/neural_compressor/quantization.py:202]:
```suggestion
                raise ValueError("Weight-only quantization needs a calibration dataset for both GPTQ and AWQ methodologies.")
```

**Review by echarlaix — COMMENTED** [review]
Thanks a lot for the addition @mengniwang95

File: examples/neural_compressor/language-modeling/run_clm.py
Comment (echarlaix) [reply examples/neural_compressor/language-modeling/run_clm.py:626]: Also could make sense to keep the same API as for the NNCF quantizer : https://github.com/huggingface/optimum-intel/blob/v1.10.1/optimum/intel/openvino/quantization.py#L115

Comment (mengniwang95) [reply examples/neural_compressor/language-modeling/run_clm.py:626]: if op_type_dict is not provided, it will use the default configuration: dtype-int4, group_size-32, scheme-sym, algorithm-RTN

File: examples/neural_compressor/language-modeling/run_clm.py
Comment (mengniwang95) [reply examples/neural_compressor/language-modeling/run_clm.py:216]: that's because if you don't set the op_type_dict, the default algorithm is 'RTN'

File: examples/neural_compressor/language-modeling/run_clm.py
Comment (mengniwang95) [reply examples/neural_compressor/language-modeling/run_clm.py:732]: sure, calculation of hessian matrix with bs>1 is WIP

File: examples/neural_compressor/language-modeling/run_clm.py
Comment (mengniwang95) [reply examples/neural_compressor/language-modeling/run_clm.py:719]: Of course we can add an enhancement in INC to move this feature to config/quantizer. This parameter is used to check calibration data length to ensure they won't exceed model's config acceptable sequence length, since we do not know how users tokenize their own datasets.

Comment (echarlaix) [reply examples/neural_compressor/language-modeling/run_clm.py:721]: Ok I see, then it could make sense to check this in the examples scripts directly, I don't think we need to add this verification in the `INCQuantizer` or in `neural-compressor` (in `GPTQuantizer`) as it belongs to a different scope in my opinion

Comment (echarlaix) [reply examples/neural_compressor/language-modeling/run_clm.py:721]: Once this point is fixed we can merge in my opinion!

File: tests/neural_compressor/test_optimization.py
Comment (mengniwang95) [reply tests/neural_compressor/test_optimization.py:185]: sure

**Review by echarlaix — APPROVED** [review]
Looks great, thanks for the addition @mengniwang95 !

**Comment by mengniwang95** [comment]
Hi Ella, currently UT fails since it doesn't use the latest master code. Do we need to wait neural-compressor 2.3 release for UT test and merge this PR after all test passing? @echarlaix

**Comment by echarlaix** [comment]
> Hi Ella, currently UT fails since it doesn't use the latest master code. Do we need to wait neural-compressor 2.3 release for UT test and merge this PR after all test passing? @echarlaix

For when is  the `neural-compressor` release planned ? This PR is compatible with the current INC latest version so I'm ok to with merging it now (the test can be added in an other PR with INC being installed from source)

**Comment by mengniwang95** [comment]
> > Hi Ella, currently UT fails since it doesn't use the latest master code. Do we need to wait neural-compressor 2.3 release for UT test and merge this PR after all test passing? @echarlaix
> 
> For when is the `neural-compressor` release planned ? This PR is compatible with the current INC latest version so I'm ok to with merging it now (the test can be added in an other PR with INC being installed from source)

Hi Ella, neural-compressor release is planned on 9/15. I add INT4 UT in this branch, but it is not triggered due to neural-compressor < 2.3

**Comment by hshen14** [comment]
@echarlaix it seems some tests failed, while they may not be related with the changes. Could you please help check, or is it okay to get this PR merged?

**Comment by echarlaix** [comment]
> @echarlaix it seems some tests failed, while they may not be related with the changes. Could you please help check, or is it okay to get this PR merged?

Could you update your branch by rebasing from main ? This will fix all unrelated tests. The INC tests are failing, because the release is previewed for tomorrow I think we should install `neural-compressor` from source [here](https://github.com/huggingface/optimum-intel/blob/v1.11.0/.github/workflows/test_inc.yml#L32-L33) to verify all the tests are passing and then we can merge

**Pull request merged**
